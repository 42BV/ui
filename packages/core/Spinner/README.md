# @42.nl/ui-core-spinner

## Installation

Install the package in your project directory with:

```sh
// with npm
npm install @42.nl/ui-core-spinner

// with yarn
yarn add @42.nl/ui-core-spinner
```
