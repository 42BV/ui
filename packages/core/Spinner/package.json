{
  "name": "@42.nl/ui-core-spinner",
  "version": "0.0.1",
  "description": "",
  "main": "lib/Spinner.js",
  "private": false,
  "scripts": {
    "test": "jest --config=jest.config.js",
    "build": "tsc",
    "lint": "eslint src --ext js,jsx,ts,tsx --max-warnings=0",
    "clean": "rm -rf lib"
  },
  "files": [
    "lib/*"
  ],
  "dependencies": {
    "classnames": "2.2.6"
  },
  "peerDependencies": {
    "react": "^16.9.0",
    "reactstrap": "^8.0.1"
  },
  "devDependencies": {
    "react": "16.9.0",
    "reactstrap": "8.0.1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  }
}
