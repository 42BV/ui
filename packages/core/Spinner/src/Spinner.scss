// @remove-on-build-start
@import '~@42.nl/ui-core-styling/lib/bootstrap';
// @remove-on-build-end

.spinner {
  animation: rotate 2s linear infinite;

  @keyframes rotate {
    100% {
      transform: rotate(360deg);
    }
  }

  .path {
    transform-origin: center;
    stroke-dasharray: 1, 200;
    stroke-dashoffset: 0;
    animation: dash 1.5s ease-in-out infinite;
    stroke-linecap: round;
  }

  @keyframes dash {
    0% {
      stroke-dasharray: 1, 200;
      stroke-dashoffset: 0;
    }
    50% {
      stroke-dasharray: 89, 200;
      stroke-dashoffset: -35;
    }
    100% {
      stroke-dasharray: 89, 200;
      stroke-dashoffset: -124;
    }
  }
}
