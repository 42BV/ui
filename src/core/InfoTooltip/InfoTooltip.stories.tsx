import React from 'react';
import { storiesOf } from '@storybook/react';

import { InfoTooltip } from './InfoTooltip';
import { Alert } from 'reactstrap';

storiesOf('core/InfoTooltip', module)
  .addParameters({ component: InfoTooltip })
  .addDecorator((Story) => (
    <>
      <Alert color="warning" className="mb-4">
        <p>To be able to use IntoTooltip, you have to add rc-tooltip to your dependencies:</p>
        <code>npm install --save rc-tooltip</code>
        <p className="mb-0 mt-2">You also have to add the stylesheet to your project</p>
        <code>@import &apos;rc-tooltip/assets/bootstrap.css&apos;;</code>
      </Alert>
      <Story />
    </>
  ))
  .add('basic', () => {
    return (
      <div className="text-center">
        <InfoTooltip tooltip="This is the content of the tooltip" />
      </div>
    );
  })
  .add('custom content', () => {
    return (
      <div className="text-center">
        <InfoTooltip
          tooltip={
            <>
              <p>
                The tooltip can contain multiple paragraphs by using plain HTML.
              </p>
              <p>
                But it is also allowed to use custom components inside the
                tooltip. Just be careful with components with interactions as
                they might conflict with the tooltip.
              </p>
            </>
          }
        />
      </div>
    );
  })
  .add('custom size', () => {
    return (
      <div className="text-center">
        <InfoTooltip tooltip="This icon is quite small" size={10} />
        <InfoTooltip tooltip="This is the default size of the icon" />
        <InfoTooltip tooltip="This icon is quite big" size={20} />
      </div>
    );
  });
