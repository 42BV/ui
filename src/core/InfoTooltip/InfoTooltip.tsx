import { Icon } from '../Icon';
import { Tooltip } from '../Tooltip/Tooltip';
import { ReactNode } from 'react';

type Props = {
  /**
   * The content of the tooltip.
   */
  tooltip: ReactNode;

  /**
   * Optionally the size of the icon.
   * Defaults to 16.
   */
  size?: number;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

/**
 * InfoTooltip is an info icon that displays a tooltip on mouseover.
 */
export function InfoTooltip(props: Props) {
  const { tooltip, size = 16, className } = props;
  return (
    <Tooltip content={tooltip} className={className}>
      <Icon icon="info" size={size} />
    </Tooltip>
  );
}
