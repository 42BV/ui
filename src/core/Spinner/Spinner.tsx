import React from 'react';
import classNames from 'classnames';

type Props = {
  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Color of the circle within the spinner.
   *
   */
  color: string;

  /**
   * Width and height of the spinner.
   *
   */
  size: number;
};

/**
 * Spinner is a component which is used as a loading indicator.
 *
 * Use it for example when data is fetching during navigation to a page.
 */
export function Spinner({ className, color, size }: Props) {
  return (
    <svg
      className={classNames('spinner', className)}
      width={size}
      height={size}
      viewBox="25 25 50 50"
    >
      <circle
        className="path"
        fill="none"
        stroke={color}
        cx="50"
        cy="50"
        r="20"
        strokeWidth="4"
        strokeMiterlimit="10"
      />
    </svg>
  );
}
