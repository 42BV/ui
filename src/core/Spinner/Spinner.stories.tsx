import React from 'react';
import { storiesOf } from '@storybook/react';

import { Spinner } from './Spinner';

storiesOf('core/Spinner', module)
  .addParameters({ component: Spinner })
  .add('default', () => {
    return (
      <div className="text-center">
        <Spinner color="" size={42} />
        <hr />
        <Spinner color="red" size={42} />
        <Spinner color="white" size={42} />
        <Spinner color="blue" size={42} />
        <hr />
        <Spinner color="orange" size={16} />
        <hr />
        <Spinner color="purple" size={500} />
      </div>
    );
  });
