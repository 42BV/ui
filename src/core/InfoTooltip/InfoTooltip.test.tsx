import React from 'react';
import { render, screen } from '@testing-library/react';

import { InfoTooltip } from './InfoTooltip';
import { Spinner } from '../Spinner/Spinner';

describe('Component: InfoTooltip', () => {
  function setup({
    tooltip = 'tooltip content',
    size,
    className
  }: {
    tooltip?: React.ReactNode;
    size?: number;
    className?: string;
  }) {
    const props = {
      tooltip,
      size,
      className
    };

    const { container } = render(<InfoTooltip {...props} />);

    return { container };
  }

  describe('ui', () => {
    test('default', () => {
      const { container } = setup({});
      expect(container).toMatchSnapshot();
    });

    test('custom content', () => {
      const { container } = setup({
        tooltip: <Spinner size={10} color="primary" />
      });
      expect(container).toMatchSnapshot();
    });

    test('with size', () => {
      setup({ size: 10 });
      expect(getComputedStyle(screen.getByText('info')).fontSize).toBe('10px');
    });
  });
});
