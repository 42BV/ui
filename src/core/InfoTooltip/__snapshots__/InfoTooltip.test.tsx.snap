// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: InfoTooltip ui custom content 1`] = `
<div>
  <span
    aria-describedby="2"
    role="button"
    style="outline: 0;"
    tabindex="0"
  >
    <i
      class="icon material-icons"
      style="font-size: 16px;"
    >
      info
    </i>
  </span>
</div>
`;

exports[`Component: InfoTooltip ui default 1`] = `
<div>
  <span
    aria-describedby="1"
    role="button"
    style="outline: 0;"
    tabindex="0"
  >
    <i
      class="icon material-icons"
      style="font-size: 16px;"
    >
      info
    </i>
  </span>
</div>
`;
